### PR TITLE
Replace LedgerCursor with getPastLedger

### PR DESCRIPTION
**Pull request by mrBliss**
Now that `getPastLedger` is cheap, we no longer need the concept of
`LedgerCursor`, so remove it entirely.

### CODE REVIEW

**Comment by mrBliss** [comment]
@nfrisby I'm going to need your help for this one.

With https://github.com/input-output-hk/ouroboros-network/pull/2459 merged, it is now cheap to ask for a past (within the last `k` blocks on the current chain) ledger. We no longer have to read blocks from disk and reapply them. So `getPastLedger` can now live in `STM m` instead of `m`.

In this PR, I make that change and also replace the concept of a `LedgerCursor` with just `getPastLedger`. You can ignore most of those changes.

Strangely enough, this PR makes some Byron ThreadNet tests fail because the `prop_svu` property is violated. 

At first I saw no connection between the change and the failure, but then I realised that we call `getPastLedger`, which is now much cheaper, when forging blocks.

I added a minimal test case for it in a separate commit (it can eventually become a regression test).

An excerpt of the trace of the minimal test case *without* the "Replace LedgerCursor with getPastLedger" commit:

```
TraceStartLeadershipCheck (SlotNo 0)
TraceMempoolAddedTx updateproposal: Dummy:0 [..]
TraceMempoolAddedTx updatevote: Update Vote { voter: 99237b83722e1ed3, proposal id: 904e41345e6e68ac } [..]
TraceNodeIsLeader (SlotNo 0)
TraceForgedBlock (SlotNo 0) <block containing the proposal and vote>
```

*With* the "Replace LedgerCursor with getPastLedger" commit:

```
TraceStartLeadershipCheck (SlotNo 0)
TraceNodeIsLeader (SlotNo 0)
TraceForgedBlock (SlotNo 0) <empty block>
TraceMempoolAddedTx updateproposal: Dummy:0 [..]
TraceMempoolAddedTx updatevote: Update Vote { voter: 99237b83722e1ed3, proposal id: 904e41345e6e68ac } [..]
```

So it seems that `forkCrucialTxs` hasn't had the chance to add the update proposal and vote to the Mempool before we forge the block. I believe it's a timing issue, as the following workaround seems to fix the failing test: add `threadDelay 0` before `getPastLedger` in `forkBlockForging`. However, I just noticed another test failing in CI :flushed:.

We should find a proper solution for this. For example, modify the `forgeTracer` to block when `TraceStartLeadershipCheck` or `TraceNodeIsLeader` is traced until we have tried to add the crucial transactions to the Mempool. You have more experience with this than I, can you help me out?

**Comment by nfrisby** [comment]
@mrBliss Oh bother, sounds tedious and brittle :( I will investigate today.

**Comment by nfrisby** [comment]
I made a commit available that updates the tests to account for your workaround (it's passed 400 RealPBFT tests). I'm not proposing this as a path towards merging this PR. It's just what I started with to re-identify the relevant concerns. I thought you might be interested, especially if you'd benefit from being at least temporarily unblocked.

```
ff1212298 - (origin/nfrisby/tmp-for-thomas-pr2543) WIP complete the workaround from the previous commit (2 minutes ago) <Nicolas Frisby>
9255f18ee - (origin/mrBliss/remove-ledgercursor) WIP add workaround (11 hours ago) <Thomas Winant>
d```

I'm now going to consider a more reliable alternative to the workaround you experimented with, as you requested.

**Comment by nfrisby** [comment]
PR 2550 derailed me, but I'm back on this now. Probably yielding something for you during my tomorrow.

**Comment by nfrisby** [comment]
@mrBliss, I think 9bed76696 adds what you what. The `*-byron-test:test:test` suite ThreadNet tests have passed 12 executions with this.

```
9bed76696 - (origin/nfrisby/remove-ledgercursor) test-infra: robustly ensure crucial txs enter mempool before block is forged (3 minutes ago) <Nicolas Frisby>
7d9b55217 - WIP add failing test case (70 minutes ago) <Thomas Winant>
```

**Comment by mrBliss** [comment]
> @mrBliss, I think [9bed766](https://github.com/input-output-hk/ouroboros-network/commit/9bed76696cebc494d9aa4819ea8aba6dbf8e2120) adds what you what. The `*-byron-test:test:test` suite ThreadNet tests have passed 12 executions with this.
> 
> ```
> 9bed76696 - (origin/nfrisby/remove-ledgercursor) test-infra: robustly ensure crucial txs enter mempool before block is forged (3 minutes ago) <Nicolas Frisby>
> 7d9b55217 - WIP add failing test case (70 minutes ago) <Thomas Winant>
> ```

Thanks for helping yet again! :slightly_smiling_face:

**Comment by mrBliss** [comment]
@nfrisby Could you review this PR? It mostly deletes code. Other than that, it simplifies the `localStateQueryServer`.

File: ouroboros-consensus/src/Ouroboros/Consensus/MiniProtocol/LocalStateQuery/Server.hs [hunk 1]
```diff
@@ -16,59 +13,49 @@ import           Ouroboros.Consensus.Ledger.Abstract
...
+            -> SendMsgAcquired $ acquired (ledgerState pastLedger)
+          Nothing
+            | pointSlot pt < pointSlot immutablePoint
+            -> SendMsgFailure AcquireFailurePointTooOld idle
+            | otherwise
+            -> SendMsgFailure AcquireFailurePointNotOnChain idle
+
+    acquired :: LedgerState blk -> ServerStAcquired blk (Query blk) m ()
+    acquired ledgerState = ServerStAcquired {
+          recvMsgQuery     = handleQuery ledgerState
```
Comment (nfrisby) [inline ouroboros-consensus/src/Ouroboros/Consensus/MiniProtocol/LocalStateQuery/Server.hs:49]: An idle question: were these message names too tied to the notion of a cursor?

Comment (mrBliss) [reply ouroboros-consensus/src/Ouroboros/Consensus/MiniProtocol/LocalStateQuery/Server.hs:49]: If I remember correctly, it's actually the other way round. The cursor was written for the `LocalStateQuery` protocol. But the `LocalStateQuery` protocol was written in mind with the LedgerDB and its (previous) cost model in mind: acquiring a snapshot was not cheap (in terms of time, but cheap in terms of space, thanks to sharing), but is now.

**Comment by mrBliss** [comment]
bors merge

**Pull request merged**